apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```diff
@@ -0,0 +1,626 @@
+            /// <param name="publisherSettings">Optional. The settings to use when creating <see cref="Publisher"/> instances.</param>
+            /// <param name="credentials">Optional. Credentials to use then creating <see cref="Publisher"/> instances.
+            /// Defaults to using the default credentials.</param>
+            public Settings(int? clientCount = null, PublisherSettings publisherSettings = null, ChannelCredentials credentials = null)
+            {
+                ClientCount = clientCount;
+                PublisherSettings = publisherSettings;
+                Credentials = credentials;
+            }
+
+            internal Settings(Settings other)
+            {
+                Clients = other.Clients != null ? new List<PublisherClient>(other.Clients) : null;
+                ClientCount = other.ClientCount;
+                PublisherSettings = other.PublisherSettings?.Clone();
+                Credentials = other.Credentials;
+                BatchingSettings = other.BatchingSettings?.Clone();
+                Scheduler = other.Scheduler;
+                BatchMaxElementCount = other.BatchMaxElementCount;
+                BatchMaxByteCount = other.BatchMaxByteCount;
+            }
+
+            /// <summary>
+            /// The clients to use within a <see cref="HiPerfPublisher"/>.
+            /// </summary>
+            public IList<PublisherClient> Clients { get; }
+
+            /// <summary>
+            /// The number of <see cref="PublisherClient"/> to create and use within a <see cref="HiPerfPublisher"/> instance.
+            /// If <c>null</c>, defaults to the CPU count on the machine this is being executed on.
+            /// </summary>
+            public int? ClientCount { get; }
+
+            /// <summary>
+            /// The settings to use when creating <see cref="Publisher"/> instances.
+            /// If <c>null</c>, defaults to <see cref="PublisherSettings.GetDefault"/>.
+            /// </summary>
+            public PublisherSettings PublisherSettings { get; }
+
+            /// <summary>
+            /// Credentials to use then creating <see cref="Publisher"/> instances.
+            /// If <c>null</c>, defaults to using the default credentials.
+            /// </summary>
+            public ChannelCredentials Credentials { get; }
+
+            /// <summary>
+            /// <see cref="BatchingSettings"/> that control how messages are batched when sending.
+            /// If <c>null</c>, defaults to <see cref="DefaultBatchingSettings"/>.
+            /// </summary>
+            public BatchingSettings BatchingSettings { get; set; }
+
+            /// <summary>
+            /// The <see cref="IScheduler"/> to use.
+            /// If <c>null</c>, defaults to <see cref="SystemScheduler"/>. Usually only useful for testing.
+            /// </summary>
+            public IScheduler Scheduler { get; set; }
+
+            /// <summary>
```

Comment:
No, this is slightly subtle, and needs better docs; or the feature removing completely.
Currently, `BatchingSettings` specifies the thresholds at which a batch is sent to the server.
However, if a batch *cannot* be sent to the server right now (because there are no available clients), then it continues filling the batch past the thresholds in `BatchingSettings`. These two following properties (`BatchMaxElementCount`, `BatchMaxByteCount`) specify the absolute maximum size it'll let a batch get to before queuing for sending.
As soon as a new batch can be sent (ie a client is available), then an 'over-threshold' batch will immediately be sent.